File: javascript/core/src/engine/verbs/chain.ts
/*!
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project.
 */
import { table } from 'arquero'

import type { StepFunction, TableStore } from '../../index.js'
import type { ChainStep, TableContainer } from '../../types.js'
import { aggregate } from './aggregate.js'
import { bin } from './bin.js'
import { binarize } from './binarize.js'
import { boolean } from './boolean.js'
import { concat } from './concat.js'
import { convert } from './convert.js'
import { dedupe } from './dedupe.js'
import { derive } from './derive.js'
import { difference } from './difference.js'
import { erase } from './erase.js'
import { fetch } from './fetch.js'
import { fill } from './fill.js'
import { filter } from './filter.js'
import { fold } from './fold.js'
import { groupby } from './groupby.js'
import { impute } from './impute.js'
import { intersect } from './intersect.js'
import { join } from './join.js'
import { lookup } from './lookup.js'
import { merge } from './merge.js'
import { orderby } from './orderby.js'
import { pivot } from './pivot.js'
import { recode } from './recode.js'
import { rename } from './rename.js'
import { rollup } from './rollup.js'
import { sample } from './sample.js'
import { select } from './select.js'
import { spread } from './spread.js'
import { unfold } from './unfold.js'
import { ungroup } from './ungroup.js'
import { union } from './union.js'
import { unorder } from './unorder.js'
import { unroll } from './unroll.js'
import { window } from './window.js'

const verbs: Record<string, StepFunction<any>> = {
	aggregate,
	bin,
	binarize,
	boolean,
	chain: exec,
	concat,
	convert,
	dedupe,
	derive,
	difference,
	erase,
	fetch,
	fill,
	filter,
	fold,
	groupby,
	impute,
	intersect,
	join,
	lookup,
	merge,
	orderby,
	pivot,
	recode,
	rename,
	rollup,
	sample,
	select,
	spread,
	unfold,
	ungroup,
	union,
	unorder,
	unroll,
	window,
}

async function exec(
	step: ChainStep,
	store: TableStore,
): Promise<TableContainer> {
	const {
		args: { steps, nofork },
	} = step

	const substore = nofork ? store : await store.clone()

	let output: TableContainer = {
		id: step.output,
		table: table({}),
	}

	for (let index = 0; index < steps.length; index++) {
		const step = steps[index]!
		const { verb } = step
		try {
			// fallback to chain if unspecified - this allows custom names to identify different chains
			const fn: StepFunction<unknown> =
				verbs[verb] || (exec as StepFunction<any>)
			// child store gets intermediate outputs so chain steps can do lookups
			output = await fn(step, substore)
			substore.set(output)
		} catch (e) {
			console.error(`Pipeline failed on step ${index}`, step)
			throw e
		}
	}
	// parent store only gets the final output of the chain
	const final = {
		...output,
		id: step.output,
	}
	store.set(final)
	// return the final table
	return final
}

/**
 * Executes a series of steps.
 * This is actually the core engine for executing pipelines.
 */
export async function chain(
	step: ChainStep,
	store: TableStore,
): Promise<TableContainer> {
	return exec(step, store)
}

File: python/data-wrangling-components/data_wrangling_components/engine/verbs/chain.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project.
#

import logging

from copy import deepcopy
from typing import List

from dataclasses import dataclass

from data_wrangling_components.engine.verbs.verbs_mapping import functions
from data_wrangling_components.table_store import TableContainer, TableStore
from data_wrangling_components.types import Step, Verb, compound_verbs


@dataclass
class ChainArgs:
    steps: List[Step]
    nofork: bool = False


def chain(step: Step, store: TableStore):
    """Runs a series of steps in a pipeline

    :param step:
        Parameters to execute the operation.
        See :py:class:`~data_wrangling_components.engine.verbs.chain.ChainArgs`
    :type step: Step
    :param store:
        Table store that contains the inputs to be used in the execution
    :type store: TableStore

    :return: the final output of executing the pipeline
    """
    args = ChainArgs(
        steps=step.args["steps"],
        nofork=step.args.get("nofork", False),
    )

    if args.nofork:
        steps_store = store
    else:
        steps_store = deepcopy(store)

    for chain_step in args.steps:
        chain_step = (
            Step(
                verb=Verb(chain_step["verb"]),
                input=chain_step["input"],
                output=chain_step["output"],
                args=chain_step["args"],
            )
            if isinstance(chain_step, dict)
            else chain_step
        )
        try:
            if chain_step.verb in compound_verbs:
                output = chain(chain_step, steps_store)
            else:
                output = functions[chain_step.verb](chain_step, steps_store)
            steps_store.set(chain_step.output, output)
        except Exception as e:
            logging.error("Pipeline failed on step '{}'", chain_step)
            raise e

    output = steps_store.table(steps_store.list()[-1])
    output_container: TableContainer = TableContainer(
        id=step.output, name=step.output, table=output
    )
    store.set(step.output, output_container)
    return output_container
